File: scripts/validate.ts
/**
 * Validate every game entry in games/ against schema/game.schema.json.
 *
 * Runs the JSON Schema check plus the cross-file rules a schema cannot express:
 * filenames matching ids, ids being unique, and player counts being internally
 * consistent. Exits non-zero if anything fails, so it works as a CI gate.
 *
 *   node scripts/validate.ts            # validate everything
 *   node scripts/validate.ts --quiet    # only print problems
 */

import { readFileSync } from "node:fs";
import { basename } from "node:path";

import { Ajv2020 } from "ajv/dist/2020.js";
import type { ErrorObject, ValidateFunction } from "ajv";

import { GAMES_DIR, SCHEMA_PATH, gameFiles } from "./games.ts";

type Entry = Record<string, unknown>;

function describe(error: ErrorObject): string {
  const location = error.instancePath.replace(/^\//, "").replace(/\//g, "/");
  return `${location || "(root)"}: ${error.message ?? "is invalid"}`;
}

/** Problems with one entry: schema errors plus the checks the schema can't do. */
function checkEntry(
  path: string,
  data: Entry,
  validate: ValidateFunction,
): string[] {
  const problems: string[] = [];

  if (!validate(data)) {
    for (const error of validate.errors ?? []) problems.push(describe(error));
  }

  const id = data["id"];
  const stem = basename(path, ".json");
  if (typeof id === "string" && id !== stem) {
    problems.push(`id "${id}" does not match filename "${basename(path)}"`);
  }

  const players = data["players"];
  if (players && typeof players === "object") {
    const { min, max, ideal } = players as Record<string, unknown>;
    if (
      typeof min === "number" &&
      typeof max === "number" &&
      typeof ideal === "number"
    ) {
      if (min > max) {
        problems.push(`players.min (${min}) is greater than players.max (${max})`);
      }
      if (ideal < min || ideal > max) {
        problems.push(
          `players.ideal (${ideal}) is outside the range ${min}-${max}`,
        );
      }
    }
  }

  return problems;
}

function main(): number {
  const quiet = process.argv.includes("--quiet");

  const schema = JSON.parse(readFileSync(SCHEMA_PATH, "utf8")) as object;
  const ajv = new Ajv2020({ allErrors: true, strict: false });
  const validate = ajv.compile(schema);

  const paths = gameFiles();
  if (paths.length === 0) {
    console.error(`No game files found in ${GAMES_DIR}`);
    return 1;
  }

  let failures = 0;
  const seenIds = new Map<string, string>();
  const seenNames = new Map<string, string>();

  for (const path of paths) {
    const name = basename(path);
    let data: Entry;
    try {
      data = JSON.parse(readFileSync(path, "utf8")) as Entry;
    } catch (error) {
      console.log(`FAIL ${name}\n  not valid JSON: ${(error as Error).message}`);
      failures += 1;
      continue;
    }

    const problems = checkEntry(path, data, validate);

    const id = data["id"];
    if (typeof id === "string") {
      const previous = seenIds.get(id);
      if (previous) problems.push(`duplicate id, also used by ${previous}`);
      else seenIds.set(id, name);
    }

    const gameName = data["name"];
    if (typeof gameName === "string") {
      const key = gameName.trim().toLowerCase();
      const previous = seenNames.get(key);
      if (previous) problems.push(`duplicate name, also used by ${previous}`);
      else seenNames.set(key, name);
    }

    if (problems.length > 0) {
      failures += 1;
      console.log(`FAIL ${name}`);
      for (const problem of problems) console.log(`  - ${problem}`);
    } else if (!quiet) {
      console.log(`ok   ${name}`);
    }
  }

  console.log(`\n${paths.length - failures}/${paths.length} entries valid.`);
  if (failures > 0) {
    console.log(`${failures} file(s) need attention.`);
    return 1;
  }
  return 0;
}

process.exit(main());

File: scripts/validate.py
#!/usr/bin/env python3
"""Validate every game entry in games/ against schema/game.schema.json.

Runs the JSON Schema check plus the cross-file rules a schema cannot express:
filenames matching ids, ids being unique, and player counts being internally
consistent. Exits non-zero if anything fails, so it works as a CI gate.

    python3 scripts/validate.py            # validate everything
    python3 scripts/validate.py --quiet    # only print problems
"""

from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path

try:
    from jsonschema import Draft202012Validator
except ImportError:
    sys.exit("jsonschema is not installed. Run: pip install -r requirements.txt")

REPO_ROOT = Path(__file__).resolve().parent.parent
GAMES_DIR = REPO_ROOT / "games"
SCHEMA_PATH = REPO_ROOT / "schema" / "game.schema.json"


def load_schema() -> dict:
    with SCHEMA_PATH.open(encoding="utf-8") as handle:
        return json.load(handle)


def iter_game_files() -> list[Path]:
    return sorted(GAMES_DIR.glob("*.json"))


def check_entry(path: Path, data: dict, validator: Draft202012Validator) -> list[str]:
    """Return a list of human-readable problems with one entry."""
    problems = []

    for error in sorted(validator.iter_errors(data), key=lambda e: list(e.path)):
        location = "/".join(str(part) for part in error.path) or "(root)"
        problems.append(f"{location}: {error.message}")

    # Cross-checks the schema itself cannot express.
    game_id = data.get("id")
    if isinstance(game_id, str) and game_id != path.stem:
        problems.append(f"id {game_id!r} does not match filename {path.name!r}")

    players = data.get("players")
    if isinstance(players, dict):
        low, high, ideal = players.get("min"), players.get("max"), players.get("ideal")
        if all(isinstance(v, int) for v in (low, high, ideal)):
            if low > high:
                problems.append(f"players.min ({low}) is greater than players.max ({high})")
            if not low <= ideal <= high:
                problems.append(
                    f"players.ideal ({ideal}) is outside the range {low}-{high}"
                )

    return problems


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--quiet", action="store_true", help="only print entries that have problems"
    )
    args = parser.parse_args()

    if not GAMES_DIR.is_dir():
        sys.exit(f"No games directory at {GAMES_DIR}")

    validator = Draft202012Validator(load_schema())
    paths = iter_game_files()
    if not paths:
        sys.exit(f"No game files found in {GAMES_DIR}")

    failures = 0
    seen_ids: dict[str, Path] = {}
    seen_names: dict[str, Path] = {}

    for path in paths:
        try:
            with path.open(encoding="utf-8") as handle:
                data = json.load(handle)
        except json.JSONDecodeError as exc:
            print(f"FAIL {path.name}\n  not valid JSON: {exc}")
            failures += 1
            continue

        problems = check_entry(path, data, validator)

        game_id = data.get("id")
        if isinstance(game_id, str):
            if game_id in seen_ids:
                problems.append(f"duplicate id, also used by {seen_ids[game_id].name}")
            else:
                seen_ids[game_id] = path

        name = data.get("name")
        if isinstance(name, str):
            key = name.strip().casefold()
            if key in seen_names:
                problems.append(f"duplicate name, also used by {seen_names[key].name}")
            else:
                seen_names[key] = path

        if problems:
            failures += 1
            print(f"FAIL {path.name}")
            for problem in problems:
                print(f"  - {problem}")
        elif not args.quiet:
            print(f"ok   {path.name}")

    print(f"\n{len(paths) - failures}/{len(paths)} entries valid.")
    if failures:
        print(f"{failures} file(s) need attention.")
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
